prophyc/generators/cpp.py

```diff
@@ -170,203 +207,169 @@ def _member_access_statement(member):
-        else:
-            members += 'return payload + 1;\n'
-        return ('template <>\n'
-                '{0}* swap<{0}>({0}* payload)\n'
-                '{{\n'
-                '{1}'
-                '}}\n').format(struct.name, _indent(members, 4))
-
-    def gen_part(part_number, part):
-        names = [mem.name for mem in part]
-        delimiters = [mem.bound for mem in part if mem.bound and mem.bound not in names]
-        members = ''.join(gen_member(mem, delimiters) + ';\n' for mem in part[:-1])
-        members += gen_last_member(struct.name + '::part{0}'.format(part_number), part[-1], delimiters)
-        return ('inline {0}::part{1}* swap({0}::part{1}* payload{3})\n'
-                '{{\n'
-                '{2}}}\n').format(struct.name,
-                                  part_number,
-                                  _indent(members, 4),
-                                  ''.join(', size_t {0}'.format(x) for x in delimiters))
-
-    main, parts = model.partition(struct.members)
-    return '\n'.join([gen_part(i + 2, part) for i, part in enumerate(parts)] +
-                     [gen_main(main, parts)])
-
-
-def _generate_swap_union(union):
-    return ('template <>\n'
-            '{0}* swap<{0}>({0}* payload)\n'
-            '{{\n'
-            '    swap(reinterpret_cast<uint32_t*>(&payload->discriminator));\n'
-            '    switch (payload->discriminator)\n'
-            '    {{\n{1}'
-            '        default: break;\n'
-            '    }}\n'
-            '    return payload + 1;\n'
-            '}}\n').format(
-                union.name,
-                ''.join(8 * ' ' + 'case {0}::discriminator_{1}: swap({2}); break;\n'.format(
-                    union.name, m.name, _member_access_statement(m)
-                ) for m in union.members))
-
-
-_generate_swap_visitor = {
-    model.Struct: _generate_swap_struct,
-    model.Union: _generate_swap_union
-}
+{content}
+}} // namespace prophy
+"""
 
+UNION_SWAP_TEMPLATE = """\
+template <>
+{name}* swap<{name}>({name}* payload)
+{{
+    swap(reinterpret_cast<uint32_t*>(&payload->discriminator));
+    switch (payload->discriminator)
+    {{\n{cases}\
+        default: break;
+    }}
+    return payload + 1;
+}}"""
+
+UNION_SWAP_CASE_TEMPLATE = """\
+        case {name}::discriminator_{member}: swap({access}); break;
+"""
 
-def _generator_swap(nodes):
-    for node in nodes:
-        fun = _generate_swap_visitor.get(type(node))
-        if fun:
-            yield fun(node)
+STRUCT_SWAP_PART_TEMPLATE = """\
+inline {name}::part{number}* swap({name}::part{number}* payload{delimiters})
+{{
+{members}}}
+"""
 
+STRUCT_SWAP_MAIN_TEMPLATE = """\
+template <>
+{name}* swap<{name}>({name}* payload)
+{{
+{members}}}"""
 
-header = """\
-#ifndef _PROPHY_GENERATED_{0}_HPP
-#define _PROPHY_GENERATED_{0}_HPP
 
-#include <prophy/prophy.hpp>
-"""
+class _CppSwapTranslator(BlockTranslatorBase):
+    block_template = SOURCE_TEMPLATE
```

Comment:
This Translator overrides both template and post_process function, base class could detect this and raise exception. I suppose only one API to post process block should be used, not both at the same time.

Reply:
It's my mistake, I forgot to remove the `block_post_process` implementation that changes nothing vs. base class. 

However I don't understand why should it be prohibited. I think if anybody wants to extend `prophyc` functionality then such person has to understand how python works and to be free to do everything that tends towards the target. Python's magic is the freedom, not the restrictions.

Reply:
OK, I checked it and really there is no need to override `block_post_process`. Everything can be done by change of the template itself. I'm turning it into semi-private (once underscored) method.

Reply:
> However I don't understand why should it be prohibited. I think if anybody wants to extend prophyc functionality then such person has to understand how python works and to be free to do everything that tends towards the target. Python's magic is the freedom, not the restrictions.

I meant: if Abc class will advertise 2 ways of block formatting, it should be clear what happens when 2 are used simultaneously by derived class - e.g. exception is raised, or both formatting mechanisms are used one after another, etc. I'm not for artificial prohibitions ;)